tests/test_modeling_common.py

```diff
@@ -1485,11 +1485,24 @@ def test_correct_missing_keys(self):
             base_model_prefix = model.base_model_prefix
 
             if hasattr(model, base_model_prefix):
+
```

Comment:
This is an important new modification.

This test tries to verify that we're correctly warning when we're saving the "core" of a model, and failing to save the head parameters, and that we yell when loading the head from this.

This test is great, and different from the newly added one.
But, since we fixed a lot of `Decoder` head which are merely a copy of the embeddings matrix, then they **are** a head, but they DO NOT have new extra parameters. hence this test would fail.

There is `self.test_missing_keys` flag to deactivate that particular test on the model testing. But it would deactivate ALL potential heads (as far as I understand), potentially opening gaps in our testing.

So I modified this test to try and detect those edge cases. I don't like putting logic like that in test as it makes reasoning hard but I failed to see another way.

This test verifies that the head does have extra parameters before actually doing the verification.

Reply:
Make sure you account in the test for upcoming models which have weights with the same name in the model with head and the base model (Niels is adding one in #19981)

Reply:
I don't see which weights are sharing names.

Seems relatively regular to me: https://github.com/huggingface/transformers/pull/19981/files#diff-a2ff7478a3e58383ae065b28d9e5e4e1e1bace51b32bc91b6eb086fabb0bd702R448-R470